### PR TITLE
[Chef] Support assigning product name when compiling Chef applications

### PR DESCRIPTION
**Pull request by erwinpan1**
Support assigning product name for chef applications. Now only `linux`, `esp32`, `nrfconnect` are supported

* Usage:
```
Usage:       usage: chef.py [options]
...
  -P PRODUCT_NAME, --pname=PRODUCT_NAME
                        specifies the Product Name. Default is TEST_PRODUCT
...
```

* Example: the "user_defined_product_name" could be assigned to the `ProductName` attribute of `BasicInformation` cluster
```
./chef.py -zcbr -P "user_defined_product_name" --use_zzz -d rootnode_onofflight_bbs1b7IaOV -t nrfconnect
```

### CODE REVIEW

File: src/include/platform/internal/GenericConfigurationManagerImpl.ipp [hunk 1]
```diff
@@ -699,6 +699,12 @@ void GenericConfigurationManagerImpl<ConfigClass>::LogDeviceConfig()
         ChipLogProgress(DeviceLayer, "  Product Id: %u (0x%X)", productId, productId);
     }
 
+    {
+        char productName[ConfigurationManager::kMaxProductNameLength + 1];
+        err = deviceInstanceInfoProvider->GetProductName(productName, sizeof(productName));
+        ChipLogProgress(DeviceLayer, "  Product Name: %s", (err == CHIP_NO_ERROR) ? productName : "(not set)");
```
Comment (andy31415) [inline src/include/platform/internal/GenericConfigurationManagerImpl.ipp:705]: Having an error here does not necessarily mean "not set".
Please log an error message containing the error text unless there is a curated list of errors that are expected if missing (e.g. KEY_NOT_FOUND or similar).

If the error would be for example `INSUFFICIENT_MEMORY`, saying 'not set' is incorrect.

Comment (erwinpan1) [reply src/include/platform/internal/GenericConfigurationManagerImpl.ipp]: @andy31415 Can you help to review this? Thanks.

File: src/include/platform/internal/GenericConfigurationManagerImpl.ipp [hunk 2]
```diff
@@ -699,6 +699,22 @@ void GenericConfigurationManagerImpl<ConfigClass>::LogDeviceConfig()
...
+        if (CHIP_NO_ERROR == err)
+        {
+            ChipLogProgress(DeviceLayer, "  Product Name: %s", productName);
+        }
+        else
+        {
+            char errorStr[160];
+            errorStr[0] = 0;
+            FormatCHIPError(errorStr, sizeof(errorStr), err);
+            ChipLogError(DeviceLayer, "  Product Name: n/a (%s)", errorStr);
```
Comment (andy31415) [inline src/include/platform/internal/GenericConfigurationManagerImpl.ipp:714]: How about using:

```
ChipLogError(DeviceLayer, "  Product Name: n/a (%" CHIP_ERROR_FORMAT ")", err.Format())
```

?

**Review by andy31415 — APPROVED** [review]

File: src/include/platform/internal/GenericConfigurationManagerImpl.ipp (hunk 2 above)
Comment (erwinpan1) [reply src/include/platform/internal/GenericConfigurationManagerImpl.ipp]: Modified accordingly. PTAL.

**Pull request merged**